README.md
# Модуль оплаты Сбербанк Интернет Эквайринг для SimplaCMS 2.*

### Особенности
* Передача корзины товаров (кассовый чек 54-ФЗ)
* Поддержка ФФД 1.05
* Если оплата не прошла или прошла с ошибкой, оплату можно повторить, хоть спустя 20 дней...
* Работает через REST
* Поддержка одноэтапной и двухэтапной оплаты


### Требования:
* php 5.6 и выше
* curl
* SimplaCMS 2.*


### Установка модуля:
1. Просто распаковать архив с файлами модуля по адресу `/payment/Sberbank/`
2. В админ-панели добавить новый способ оплаты `Сбербанк-Эквайринг`
3. Настроить модуль в соответствии с вашими требованиями

### Настройка:
* `После оплаты сменить статус заказа на (ID статуса)` - если вы хотите что бы **после оплаты у заказа менялся статус**, укажите ID статуса заказа.  
Например `Новый` имеет id `0`, `Приняты` id `1`, `Выполнены` id - `2` итд...
* `режим оплаты (одностадийный)(по умолчанию)` - блокирование и списание средств происходит в один этап. Этот вид платежей предпочтительней, если товар или услуга предоставляется сразу после оплаты.
* `Режим оплаты (двухстадийный)` - `Внимание! обратитесь в банк для включения Двухстадийной оплаты.`Двухстадийные платежи следует использовать, если между решением покупателя произвести оплату и поставкой выбранного товара или услуги проходит какое-то время. 
Оплата производится в два этапа. На первом этапе происходит проверка наличия и блокирование средств плательщика (пре-авторизация); далее, на втором этапе, компания либо подтверждает необходимость списания средств, либо отменяет блокировку средств.

* `Передавать данные для печати чека (54-ФЗ)` - товары в заказе. Так-же нужно включить ФФД 1.05 в Кассе и Сбербанк-админке

* `Стоимость доставки включить`:  
`в стоимость товаров` - Цена доставки равномерно "размазывается" по каждой позиции в чеке, если в настройках выбранного метода доставки НЕ установлена галочка `Оплачивается отдельно`  
`как отдельную позицию в чеке` - тут все понятно. Должны быть коды оквэд у вашей организации.

* `CSS класс кнопки "перейти к оплате"` - если есть необходимость установить доп. классы css для кнопки "перейти к оплате", например 

### Установка "Разный НДС для товаров":
1. Копируем файл `/payment/Sberbank/_sql_taxType_install.php` в корень сайта
2. Идем по адресу `http://ИМЯ_ВАШЕГО_САЙТА/_sql_taxType_install.php`
3. Открываем файл `/api/Products.php`
4. Ищем функции `get_products` и `get_product`
5. Ищем в этих функциях `p.visible,` и после добавляем `p.taxType,`
6. Открываем файл `/simpla/design/html/product.tpl`, ищем `Параметры страницы`
7. Добавляем в этот блок: 
    ```
    <select name="taxType">
        <option value='0' {if $product->taxType=='0'}selected{/if}>без НДС</option>
        <option value='1' {if $product->taxType=='1'}selected{/if}>НДС по ставке 0%</option>
        <option value='2' {if $product->taxType=='2'}selected{/if}>НДС чека по ставке 10%</option>
        <option value='3' {if $product->taxType=='3'}selected{/if}>НДС чека по ставке 18%</option>
        <option value='4' {if $product->taxType=='4'}selected{/if}>НДС чека по расчетной ставке 10/110</option>
        <option value='5' {if $product->taxType=='5'}selected{/if}>НДС чека по расчетной ставке 18/118</option>
        <option value='6' {if $product->taxType=='6'}selected{/if}>НДС чека по ставке 20%</option>
        <option value='7' {if $product->taxType=='7'}selected{/if}>НДС чека по расчётной ставке 20/120</option>
    </select>
    ```
8. Открываем `/simpla/ProductAdmin.php`
9. Ищем `$product->visible = $this->request->post('visible', 'boolean');` и после добавляем `$product->taxType = $this->request->post('taxType', 'integer');`
10. В настройках платежного модуля, в пункте `Разный НДС у товаров? `, ставим `да`.


Офф. документация:
-
* https://securepayments.sberbank.ru/wiki/doku.php/integration:api:rest:requests:register_cart#items

Тестовые карты:
-
* https://developer.sberbank.ru/doc/v1/acquiring/rest-test-cards

Тестовая среда:
-
* ЛК оператора https://3dsec.sberbank.ru/mportal3   
* API: `storename-api`
* Оператор: `storename-operator`
* Пароль на оба логина: `storename`


changelog.md
v 1.6.4 - 05.06.2020
* (bugfix) - Фикс ошибки `Несоответствие между [orderBundle.cartItems.item.itemPrice] и [orderBundle.cartItems.item.itemAmount]`. Пофиксил через ceil валюты.
* (upd) - частичный рефакторинг
* (bugfix) - Если доставка была отдельной позицией в чеке, заказ не продолжался

v 1.6.3 - 29.04.2020
* (add) Добавлена опция `sbr_order_status`. В неё нужно устанавливать статус заказа, если вы хотите что бы после оплаты у заказа менялся статус.
* (add) Добавлена опция `sbr_css_checkout_button`. Создана для установки css классов на кнопку `checkout_button` "перейти к оплате".

v 1.6.2 - 27.04.2020
* (add) - Добавлен хеш заказа. Теперь, если у заказа были изменены парметры `delivery_price - цена доставки`, `total_price - общая цена заказа`, `separate_delivery - оплата за доставку оплачивается отдельно`, `coupon_discount - купон пользователя`, `discount - скидка`, будет сбрасываться хеш.  
Если хеш сброшен, значит урл заказа пересоздается в ПШ.
* (bugfix) Полностью переписана функция подгоняющая стоимость товаров в чеке, кроме доставки, к общей цене заказа. 

v 1.6.1 - 15.04.2020
* (bugfix) опять проблема копеек, когда опция `Стоимость доставки включить в стоимость товаров` "размазывалась" по товарам не совсем корректно, до копеек. Если сейчас будет какое-то расхождение, сумма расхождения будет плюсоваться к последней позиции в чеке.

v 1.6 - 28.04.2019
* (edit) Мелкие исправления орфографии
* (edit) Дата истечения заказа 20 дней. После этого времени заказ будет пересоздан с префиксом.
* (edit) Крупный рефакторинг `Sberbank.php`
* (bugfix) Не учитывался параметр способа доставки `опачивается отдельно`, при опции `Стоимость доставки включить в стоимость товаров`
* (bugfix) Доставка, как отдельная позиция в чеке, прописывалась как товар, а не услуга.

v 1.5.3 - 19.03.2019
* (add) Опция `Режим оплаты`. Теперь можно производить Двухстадийные платежи. 

v 1.5.2 - 22.01.2019
* (add) Получение информации о чеке и запись их в payment_details (в данный момент невозможно использовать в callback.php, т.к. возникает задержка получения данных от АТОЛ Сбером)
* (bugfix) Передавался неверный taxType, если использовался "Разный НДС у товаров"

v 1.5.1
* (edit) Переработана документация по установке и настройке модуля

v 1.5
* (add) Стоимость доставки можно зашить в стоимость товаров
* (edit) Правка ФФД 1.05 параметров
* (add) Добавлены новые значения taxType касательно 20% НДС
* (edit) Дата оплаты в callback.php

v1.4
* (add) Поддержка ФФД 1.05, itemAttributes (только - paymentMethod и paymentObject). 
ВАЖНО! Нужно включить поддержку в Сбербанк-Админке и Кассе (например Атол)
А после включить настройку в админ. панели сайта
* (bugfix) Проблема с округлением копеек. Теперь если стоимость не
точная, создается несколько позиций из одной. 1 - с откорректированной
ценой, 2 - с оригинальной. т.е. купив 3 "булка хлеба" по 2.33 $, и применив
скидку 5% мы получим некорректные цены. 
Общая корзина без скидки - 6.99;
Корзина со скидкой - 6.6405 (округлив round - 6.64);
т.к. нам нужно из каждого товара вычесть скидку (таково правило скидок в фз-54 на чеках),
цена товара - 2,2135;
Округлив через round - 2,21;
2.21 * на 3 товара = 6,63;
Итого у нас расхождение 0.01 цент.
Создаем дубликат товара в позиции - amount - 1, price = 2.22
B и дописываем старые товары - amount - 2, price = 2.21. 
В сумме получится что надо.
см. подробнее - https://help.dreamkas.ru/hc/ru/articles/360000185225-%D0%A1%D0%BA%D0%B8%D0%B4%D0%BA%D0%B8-%D0%B8-%D1%80%D0%B0%D0%B7%D0%B4%D0%B5%D0%BB%D0%B5%D0%BD%D0%B8%D0%B5-%D0%BF%D0%BE%D0%B7%D0%B8%D1%86%D0%B8%D0%B9-%D0%B2-%D1%87%D0%B5%D0%BA%D0%B5-%D0%BF%D0%BE-54-%D0%A4%D0%97


v1.3
* (add) Установка кастомного НДС у товаров. Инструкция по установке - /payments/Sberbank/taxType_install/инструкция.txt
* (bugfix) Проблема с округлением копеек. Для сбера важна точная стоимость.
* (bugfix) Проблема с callback стоимости товара с копейками. Был Incorrect price

v1.2
* (add) Функция передачи корзины для ФЗ-54
* (edit) Рефакторинг кода
* (edit) Переделана функция продолжения оплаты заказа

RBS.php
<?php

class RBS
{
    /**
     * АДРЕС ТЕСТОВОГО ШЛЮЗА
     *
     * @var string
     */
    const TEST_URL = 'https://3dsec.sberbank.ru/payment/rest/';

    /**
     * АДРЕС БОЕВОГО ШЛЮЗА
     *
     * @var string
     */
    const PROD_URL = 'https://securepayments.sberbank.ru/payment/rest/';

    /**
     * ЛОГИН МЕРЧАНТА
     *
     * @var string
     */
    private $user_name;

    /**
     * ПАРОЛЬ МЕРЧАНТА
     *
     * @var string
     */
    private $password;

    /**
     * ДВУХСТАДИЙНЫЙ ПЛАТЕЖ
     *
     * Если значение true - будет производиться двухстадийный платеж
     *
     * @var boolean
     */
    private $two_stage;

    /**
     * ТЕСТОВЫЙ РЕЖИМ
     *
     * Если значение true - плагин будет работать в тестовом режиме
     *
     * @var boolean
     */
    private $test_mode;


    /**
     * КОНСТРУКТОР КЛАССА
     *
     * Заполнение свойств объекта
     *
     * @param string $user_name логин мерчанта
     * @param string $password пароль мерчанта
     * @param boolean $two_stage двухстадийный платеж
     * @param boolean $test_mode тестовый режим
     */
    public function __construct($user_name, $password, $two_stage, $test_mode)
    {
        $this->user_name = $user_name;
        $this->password = $password;
        $this->two_stage = $two_stage;
        $this->test_mode = $test_mode;
    }

    /**
     * ЗАПРОС В ПШ
     *
     * Формирование запроса в платежный шлюз и парсинг JSON-ответа
     *
     * @param string $method метод запроса в ПШ
     * @param string[] $data данные в запросе
     * @param string $url адрес ПШ
     * @return string[]
     */
    private function gateway($method, $data)
    {
        $data['userName'] = $this->user_name;
        $data['password'] = $this->password;
        if ($this->test_mode) {
            $url = self::TEST_URL;
        } else {
            $url = self::PROD_URL;
        }

        $curl = curl_init();
        curl_setopt_array($curl, [
            CURLOPT_URL => $url . $method,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_POST => true,
            CURLOPT_POSTFIELDS => http_build_query($data)
        ]);
        $response = curl_exec($curl);
        $response = json_decode($response, true);
        curl_close($curl);
        return $response;
    }


    /**
     * РЕГИСТРАЦИЯ ЗАКАЗА
     * Метод register.do или registerPreAuth.do
     * @param string $order_number номер заказа в магазине
     * @param integer $amount сумма заказа
     * @param string $return_url страница, на которую необходимо вернуть пользователя если платеж прошел успешно
     * @param string $order_description Описание заказа
     * @param array $order_bundle Корзина товаров заказа
     * @param array $params Список доп. параметров (expirationDate, taxSystem)
     * @param integer $taxSystem Система налогообложения
     * @return string[]
     */
    function register_order($order_number, $amount, $return_url, $order_description, $order_bundle, $params = [])
    {
        $data = [
            'orderNumber' => $order_number,
            'amount' => $amount,
            'returnUrl' => $return_url,
            'description' => $order_description,
            'orderBundle' => $order_bundle
        ];

        // Запишем доп. параметры
        if (!empty($params)) {
            foreach ($params as $param_name => $val) {
                if($val !== false) {
                    $data[$param_name] = $val;
                }
            }
        }

        if ($this->two_stage) {
            $method = 'registerPreAuth.do';
        } else {
            $method = 'register.do';
        }
        $response = $this->gateway($method, $data);
        return $response;
    }

    /**
     * СТАТУС ЗАКАЗА ПО ORDER ID в мерчанте
     *
     * Метод getOrderStatusExtended.do
     *
     * @param string $orderId номер заказа
     * @return string[]
     */
    public function get_order_status_by_orderId($orderId)
    {
        $data = ['orderId' => $orderId];
        $response = $this->gateway('getOrderStatusExtended.do', $data);
        return $response;
    }

    /**
     * СТАТУС ЗАКАЗА ПО ORDER NUMBER в магазине
     *
     * Метод getOrderStatusExtended.do
     *
     * @param string $order_number номер заказа в магазине
     * @return string[]
     */
    public function get_order_status_by_orderNumber($order_number)
    {
        $data = ['orderNumber' => $order_number];
        $response = $this->gateway('getOrderStatusExtended.do', $data);
        return $response;
    }


    /**
     * Запрос отмены заказа
     * Метод reverse.do
     *
     * @param string $orderId номер заказа в мерчанте
     * @return string[]
     */
    public function reverse_order($orderId)
    {
        $data = ['orderId' => $orderId];
        $response = $this->gateway('reverse.do', $data);
        return $response;
    }


    /**
     * Запрос сведений о кассовом чеке по НОМЕРУ ЗАКАЗА В МАГАЗИНЕ
     *
     * Метод getReceiptStatus.do
     *
     * @param string $order_number номер заказа в магазине
     * @return string[]
     */
    public function get_receipt_status_by_orderNumber($order_number)
    {
        $data = ['orderNumber' => $order_number];
        $response = $this->gateway('getReceiptStatus.do', $data);
        return $response;
    }


    /**
     * Запрос сведений о кассовом чеке по НОМЕРУ ЗАКАЗА В МЕРЧАНТЕ
     *
     * Метод getReceiptStatus.do
     *
     * @param string $order_id номер заказа в мерчанте
     * @return string[]
     */
    public function get_receipt_status_by_orderId($order_id)
    {
        $data = ['orderId' => $order_id];
        $response = $this->gateway('getReceiptStatus.do', $data);
        return $response;
    }


    /**
     * MDORDER ИЗ FORMURL
     *
     * @param string $url адрес платежной страницы
     * @return string
     */
    public function get_mdOrder_from_url($url)
    {
        $parse = parse_url($url);
        $mdOrder = explode('=', $parse['query']);
        return $mdOrder[1];
    }


    /**
     * СПИСОК РАЗРЕШЕННЫХ КОДОВ
     *
     * Коды, при которых можно продолжить оплату заказа в ПШ
     *
     * @param $actionCode
     * @return bool
     */
    public function allowed_actionCode($actionCode) {

        $allowed_actionCode = [
            '-100', // Не было попыток оплаты.
            // '-102', // Платеж отменен платежным агентом.
            '0', // Платеж успешно прошел.
            // '-2007' // Истек срок ожидания ввода данных.
        ];

        if (in_array($actionCode, $allowed_actionCode)) {
            return true;
        } else {
            return false;
        }

    }

}

callback.php
<?php

// Работаем в корневой директории
chdir('../../');
require_once('api/Simpla.php');
require_once('payment/Sberbank/RBS.php');

$simpla = new Simpla();

$order = $simpla->orders->get_order(intval($simpla->request->get('order', 'integer')));

if (empty($order)) {
    die('Оплачиваемый заказ не найден');
}

$method = $simpla->payment->get_payment_method(intval($order->payment_method_id));
if (empty($method)) {
    die("Неизвестный метод оплаты");
}

$settings = unserialize($method->settings);
$payment_currency = $simpla->money->get_currency(intval($method->currency_id));

/**
 * Проверим статус заказа
 */
$rbs = new RBS($settings['sbr_login'], $settings['sbr_password'], $settings['two_stage'] ? TRUE : FALSE, $settings['sbr_mode'] ? TRUE : FALSE);
$order_id_merchant = $simpla->request->get('orderId', 'string');
$response = $rbs->get_order_status_by_orderId($order_id_merchant);

// Если указана ошибка оплаты
if ($response['errorCode']) {
    die($response['errorMessage']);
}

if ($response['actionCode'] != 0) {
    header('Location: ' . $simpla->config->root_url . '/order/' . $order->url);
    die("Ошибка оплаты. " . $response['actionCodeDescription']);
}

// Нельзя оплатить уже оплаченный заказ  
if ($order->paid) {
    header('Location: ' . $simpla->config->root_url . '/order/' . $order->url, true, 302);
    exit();
}

// Проверяем оплаченный заказ
$total_price = round($order->total_price, 2) * 100;
if ($response['amount'] != (int)$total_price || $response['amount'] <= 0) {
    die("incorrect price");
}

// Получим данные о чеке
$payment_details = $rbs->get_receipt_status_by_orderId($order_id_merchant);

// Данные по заказу, установим на "оплачен" и заполним другие данные
$order_update = [
    'paid' => 1,
    'payment_date' => date('Y-m-d H:i:s'),
    'payment_details' => json_encode([
        'orderId' => $payment_details['orderId'], // Номер заказа в Платежном Шлюзе
        'uuid' => $payment_details['receipt'][0]['uuid'], // Номер чека
    ])
];

// Сменим статус заказа после оплаты
if ($settings['sbr_order_status']) {
    $order_update['status'] = $settings['sbr_order_status'];
}

// Установим флаг "оплачен"
$simpla->orders->update_order(intval($order->id), $order_update);
// $simpla->orders->pay(intval($order->id)); // Должно быть так, но не работает

// Спишем товары
$simpla->orders->close(intval($order->id));

// Отправим уведомление на email
$simpla->notify->email_order_user(intval($order->id));
$simpla->notify->email_order_admin(intval($order->id));

header('Location: ' . $simpla->config->root_url . '/order/' . $order->url, true, 302);
exit();

Sberbank.php
<?php

require_once 'api/Simpla.php';
require_once 'RBS.php';

class Sberbank extends Simpla
{
    private $payment_method = [],
        $order = [],
        $payment_settings = [],
        $debug = 0;

    public function checkout_form($order_id, $button_text = null)
    {
        if (empty($button_text)) {
            $button_text = 'Перейти к оплате';
        }

        $this->order = $this->orders->get_order((int)$order_id);
        $this->payment_method = $this->payment->get_payment_method($this->order->payment_method_id);
        $this->payment_settings = $this->payment->get_payment_settings($this->payment_method->id);

        // Получаем payment_details отдельно, т.к. в get_order он не берется.
        $this->db->query($this->db->placehold("SELECT payment_details FROM __orders WHERE id=? LIMIT 1", $this->order->id));
        $payment_details = $this->db->result();
        $this->order->payment_details = json_decode($payment_details->payment_details, TRUE);

        // Debug mode
        $this->payment_settings['sbr_debug'] == 1 AND $_SESSION['admin'] ? $this->debug = 1 : $this->debug = 0;

        $return_url = $this->config->root_url . "/payment/Sberbank/callback.php?order=" . $this->order->id;
        $order_description = 'Оплата заказа №' . $this->order->id . ' на сайте ' . $this->settings->site_name;
        $order_description = (string)mb_substr($order_description, 0, 99);

        /**
         * Подключаемся к эквайрингу
         */
        $rbs = new RBS($this->payment_settings['sbr_login'], $this->payment_settings['sbr_password'], $this->payment_settings['two_stage'] ? TRUE : FALSE, $this->payment_settings['sbr_mode'] ? TRUE : FALSE);


        /**
         * Подготавливаем корзину для 54-ФЗ
         */
        $orderBundle = [];
        if ($this->payment_settings['sbr_orderBundle']) {

            // Информация о пользователе
            $orderBundle['customerDetails'] = [
                "email" => $this->order->email,
                "phone" => preg_replace('/[^0-9]/', '', $this->order->phone)
            ];

            // Добавляем товары в чек
            $purchases = $this->normalize($this->orders->get_purchases(['order_id' => $this->order->id]));
            foreach ($purchases as $key => $purchase) {
                $orderBundle['cartItems']['items'][$key] = [
                    "positionId" => $key + 1,
                    "name" => $purchase->product_name,
                    "quantity" => [
                        "value" => $purchase->amount,
                        "measure" => $this->settings->units
                    ],
                    "itemAmount" => $purchase->price * $purchase->amount,
                    "itemCode" => $purchase->variant_id . '-'. $key,
                    "tax" => [
                        "taxType" => isset($purchase->taxType) ? $purchase->taxType : $this->payment_settings['sbr_taxType']
                    ],
                    "itemPrice" => $purchase->price,

                    // ФФД 1.05
                    "itemAttributes" => [
                        'attributes' => [
                            [
                                'name' => 'paymentMethod',
                                'value' => $this->payment_settings['sbr_paymentMethod']
                            ],
                            [
                                'name' => 'paymentObject',
                                'value' => $this->payment_settings['sbr_paymentObject']
                            ]
                        ]
                    ]
                ];
            }

            // Добавляем доставку в чек
            if ($this->payment_settings['sbr_delivery'] == 'item' AND $this->order->delivery_id AND $this->order->delivery_price > 0 AND !$this->order->separate_delivery) {

                $delivery = $this->delivery->get_delivery($this->order->delivery_id);

                $key = count($orderBundle['cartItems']['items']);
                $orderBundle['cartItems']['items'][$key] = [
                    "positionId" => count($orderBundle['cartItems']['items']) + 1,
                    "name" => $delivery->name,
                    "quantity" => [
                        "value" => 1,
                        "measure" => 'шт'
                    ],
                    "itemAmount" => $this->convert_price($this->order->delivery_price),
                    "itemCode" => 'DELIVERY-' . $delivery->id,
                    "tax" => [
                        "taxType" => $this->payment_settings['sbr_taxType']
                    ],
                    "itemPrice" => $this->convert_price($this->order->delivery_price),

                    // ФФД 1.05
                    "itemAttributes" => [
                        'attributes' => [
                            [
                                'name' => 'paymentMethod',
                                'value' => $this->payment_settings['sbr_paymentMethod']
                            ],
                            [
                                'name' => 'paymentObject',
                                'value' => 4
                            ]
                        ]
                    ]
                ];
            }

            $orderBundle = json_encode($orderBundle);
        }


        /**
         * узнаем hash заказа
         * на случай если в заказе что-то менялось, а урл на оплату еже сгенерирован и есть в payment_details
         */
        $order_token_text = $this->order->id . $this->order->delivery_price . $this->order->total_price . $this->order->separate_delivery . $this->order->coupon_discount . $this->order->discount;
        $order_token = false; // Совпадают ли хеши?
        if (isset($this->order->payment_details['order_token']) AND $this->config->check_token($order_token_text, $this->order->payment_details['order_token'])) {
            $order_token = true;
        }

        // Узнаем статус заказа в ПШ
        if (isset($this->order->payment_details['orderId'])) {
            $order_status = $rbs->get_order_status_by_orderId($this->order->payment_details['orderId']);
        } else {
            $order_status = $rbs->get_order_status_by_orderNumber($this->order->id);
        }

        // Истекло время заказа в ПШ?
        $order_expiration = strtotime($this->order->payment_details['expirationDate']) > strtotime('now') ? true : false;


        /**
         * Заказ существует, не оплачен, можно оплатить
         */
        $result = '';
        if ($order_token AND isset($order_status['actionCode']) AND $rbs->allowed_actionCode($order_status['actionCode']) AND $order_expiration) {
            $result = '<a href="' . $this->order->payment_details['formUrl'] . '" class="checkout_button ' . $this->payment_settings['sbr_css_checkout_button'] . '">' . $button_text . ' </a>';
        } elseif (!$order_token OR $order_status['errorCode']==6 OR !isset($order_status['actionCode']) OR !$rbs->allowed_actionCode($order_status['actionCode'])) {

            /**
             * Заказ не создавался ИЛИ вернули плохой код, пересоздаем
             */
            $order_prefix = (!$order_token OR (isset($order_status['actionCode']) AND !$rbs->allowed_actionCode($order_status['actionCode']))) ? '-' . strtotime(date('Y/m/d H:i:s')) : '';

            /**
             * Создаем новый заказ в ПШ
             */
            // Дата истечения оплаты через 20 дней
            //$datetime = new DateTime(date("Y-m-d H:i:s", strtotime("+1 minutes")));
            $datetime = new DateTime(date("Y-m-d H:i:s", strtotime("+20 days")));
            $this->order->payment_details['expirationDate'] = $datetime->format(DateTime::ATOM);

            // Заказ не был создан, создаем.
            $response = $rbs->register_order(
                $this->order->id . $order_prefix,
                $this->convert_price($this->order->total_price),
                $return_url,
                $order_description,
                $orderBundle,
                [
                    'expirationDate' => $this->order->payment_details['expirationDate'],
                    'taxSystem' => $this->payment_settings['sbr_taxSystem']
                ]
            );

            if (!$response['errorCode']) {

                $this->order->payment_details['orderId'] = $response['orderId'];
                $this->order->payment_details['formUrl'] = $response['formUrl'];
                $this->order->payment_details['orderNumber'] = $response['orderNumber'];
                $this->order->payment_details['order_token'] = $this->config->token($order_token_text);

                // Обновим payment_details в заказе
                $this->orders->update_order($this->order->id, ['payment_details' => json_encode($this->order->payment_details)]);

                $result = '<a href="' . $this->order->payment_details['formUrl'] . '" class="checkout_button ' . $this->payment_settings['sbr_css_checkout_button'] . '">' . $button_text . ' </a>';
            } elseif ($response == NULL) {
                $result = '<p class="checkout_info">Невозможно подключиться к платежному шлюзу</p>';
            } else {
                $result = '<p class="checkout_info">' . $response['errorMessage'] . '</p>';
            }

        } elseif ($order_status['actionCode'] == '0') {
            // Заказ был оплачен
            $result = '<p class="checkout_info">Заказ был оплачен.</p>';
        } else {
            $result = '<p class="checkout_info">' . $order_status['actionCode'] . '</p>';
        }


        if ($this->debug) {
            print '<pre>';
            echo '<h3>Статус заказа от сбера:</h3>';
            var_dump($order_status);
            echo '<h3>Сумма заказа для Сбера:</h3>';
            var_dump($this->convert_price($this->order->total_price));
            echo '<h3>Текущий заказ в БД:</h3>';
            var_dump($this->order);
            echo '<h3>Корзина для ФЗ-54:</h3>';
            var_dump($orderBundle);
            var_dump(json_decode($orderBundle));
            print '</pre>';
        }

        return $result;
    }

    /**
     * Подгоняет стоимость товаров в чеке, кроме доставки, к общей цене заказа
     *
     * @param object $purchases товары в заказе
     * @return object $purchases
     */
    private function normalize($purchases)
    {
        // Общая стоимость заказа (с учетом процентной скидки)
        $total_price = $this->order->total_price;

        // Если есть доставка, отнимаем стоимость доставки от общей суммы заказа
        if ($this->order->delivery_price && $this->order->delivery_price > 0 && !$this->order->separate_delivery) {
            $total_price -= $this->order->delivery_price;
        }

        // Добавляем стоимость скидки coupon_discount
        $total_price += $this->order->coupon_discount;

        $items_total_price = 0; // Общая стоимость позиций по отдельности.
        $corrected_purchases = []; // Корректирующий массив цен
        foreach ($purchases as $key => $item) {
            $items_total_price += $item->price * $item->amount;
            $corrected_purchases[$key] = $item->price;
        }

        /**
         * размазываем discount на все товары
         * discount - процентная скидка
         */
        if ($this->order->discount > 0) {

            // объем процентной скидки - вычитаю процент скидки / пользовательской скидки
            $discount_sum = ($items_total_price * $this->order->discount) / 100;

            foreach ($purchases as $key => $item) {
                $corrected_purchases[$key] -= $this->coefficient_price($item->amount, $item->price, $discount_sum, $items_total_price);

                if ($this->debug) {
                    echo 'Цена за item с учетом процентной скидки: ' . $corrected_purchases[$key] . '<br>';
                }
            }
        }

        /**
         * размазываем coupon_discount на все товары
         * coupon_discount - скидка по купону
         */
        if ($this->order->coupon_discount > 0) {
            foreach ($purchases as $key => $item) {
                // Вычислим процентное соотношение item price * amount от общей суммы заказа
                $corrected_purchases[$key] -= $this->coefficient_price($item->amount, $item->price, $this->order->coupon_discount, $items_total_price);

                if ($this->debug) {
                    echo 'Цена за item с учетом скидки по купону: ' . $corrected_purchases[$key] . '<br>';
                }
            }
        }


        /*
         * DELIVERY
         * Добавляем доставку в каждый товар
        */
        if ($this->payment_settings['sbr_delivery'] == 'include_item' AND !$this->order->separate_delivery) {
            foreach ($purchases as $key => $item) {
                $corrected_purchases[$key] += $this->coefficient_price($item->amount, $item->price, $this->order->delivery_price, $items_total_price);

                if ($this->debug) {
                    echo 'Цена за item с учетом доставки: ' . $corrected_purchases[$key] . '<br>';
                }
            }
        }


        /*
         * Смотрим финальную разницу, если она есть, добавляем к последней позиции разницу.
         */
        $all_sum = 0;
        $all_sum_diff = 0;
        foreach ($purchases as $key => $item) {
            $item->price = $corrected_purchases[$key];
            $all_sum += $item->price * $item->amount;
        }

        // Если доставка как отдельная позиция
        if($this->payment_settings['sbr_delivery'] == 'item' && $this->order->delivery_price && $this->order->delivery_price > 0 && !$this->order->separate_delivery) {
            $all_sum += $this->order->delivery_price;
        }

        // Корректируем
        if ($this->order->total_price != $all_sum) {
            $all_sum_diff = $this->order->total_price - $all_sum;
            $all_sum_diff = round($all_sum_diff, 2);

            // or `array_key_last` if php 7 >= 7.3.0
            end($purchases);
            $last_key = key($purchases);

            // Дублируем последнюю позицию в чеке, если их несколько для правильной корректировки цен
            if ($purchases[$last_key]->amount > 1) {
                $purchases[$last_key]->amount -= 1;
                $copy_item = clone $purchases[$last_key];
                $copy_item->amount = 1;
                $copy_item->price += $all_sum_diff;
                $purchases[] = $copy_item;
            } else {
                $purchases[$last_key]->price += $all_sum_diff / $purchases[$last_key]->amount;
            }

            if($this->debug) {
                echo '<h3>Корректировка сумы заказа</h3>';
                echo '$this->order->total_price: ' . $this->order->total_price . '<br/>';
                echo '$all_sum: ' . $all_sum . '<br/>';
                echo 'Отличие от суммы заказа: ' . $all_sum_diff . '<br/>';
            }
        }

        /*
         * Формируем цены для сбера
         */
        foreach ($purchases as $item) {
            $item->price = $this->convert_price($item->price);
        }

        /**
         * Кастомный НДС
         * для каждого товара
         */
        if ($this->payment_settings['sbr_taxProduct']) {
            $products_ids = [];
            foreach ($purchases as $item) {
                $products_ids[] = $item->product_id;
            }
            $products = [];
            foreach ($this->products->get_products(['id' => $products_ids]) as $product) {
                $products[$product->id] = $product;
            }
            foreach ($purchases as $item) {
                if (isset($products[$item->product_id]->taxType)) {
                    $item->taxType = $products[$item->product_id]->taxType;
                }
            }
        }

        return $purchases;
    }

    /**
     * Конвертируем цену из 100.00 в 10000 для Платежного Шлюза
     *
     * @param $price
     * @return integer
     */
    private function convert_price($price)
    {
        // return $this->money->convert($price, $this->payment_method->currency_id, false) * 100;
        return round($price, 2) * 100;
    }

    /**
     * Считаем коэффициент
     *
     * @param $item_amount
     * @param $item_price
     * @param $sum
     * @param $items_total_price
     * @return false|float
     */
    private function coefficient_price($item_amount, $item_price, $sum, $items_total_price) {
        $coefficient = ($item_amount * $item_price) * 100 / $items_total_price;
        return ceil(round((($sum * $coefficient) / 100) / $item_amount, 2));
    }
}